be/src/storage/rowset/bloom_filter_index_writer.cpp

```diff
@@ -226,9 +226,11 @@ class NgramBloomFilterIndexWriterImpl<field_type, std::enable_if_t<is_slice_type
                     if (this->_bf_options.case_sensitive) {
                         _values.insert(get_value<field_type>(&cur_ngram, this->_typeinfo, &this->_pool));
                     } else {
-                        // todo::exist two copy of ngram, need to optimize
+                        // TODO: exist two copies of ngram, need to optimize
+                        // Use UTF-8 aware tolower for proper Unicode case folding
                         std::string lower_ngram;
-                        Slice lower_ngram_slice = cur_ngram.tolower(lower_ngram);
+                        utf8_tolower(cur_ngram.get_data(), cur_ngram.get_size(), lower_ngram);
```

Comment:
utf8_tolower costs more time than its ASCII's version, so we should check the string if it is an ASCII string or not, if so, we use ASCII to_lower instead.

slice_gram_num from LN#215 can used to judge if the cur_slice is an ASCII string.  if slice_gram_num == cur_slice.size, then it is.